### ISSUE TITLE
File Manager: Future improvements

### ISSUE THREAD

**Issue by qooob** [qooob/authentic-theme#629]
To make new File Manager to cover all of the features of old Java File Manager we need to add:

 1. Mount feature

 2. Manage symbolic links

 3. Extend search with ability to specify user and group, file type (file, directory, symlink)

To make it more powerful and comfortable to use, lets add some extra functionality:
 1. Ability to diff two files #612 
 2. Movable and resizeable editor windows
 3. Image viewer with simple editor functionality; audio and video player
 4. Viewing archives content
 5. Ability to downloaddirectory
6. Selecting files/directories using wildcard
7. Calculate and verify checksum of files
8. Split and combine files
9. Exclud from search given folders

10. Multi-rename tool

Anything else you miss/wish? Please comment below.

**Comment by qooob** [comment]
11. Add multiple uploads
12. Add an option to automatically extract uploaded file(s)

**Comment by campones** [comment]
it would be nice to be able to reproduce the tree of directories on the left side, because it allows a quick access.  directories previously opened stay on that state even after retrieving another directory

**Comment by campones** [comment]
it looks good but I don't think it gives a good global view about the file on the system

maybe I  just need to get used to

**Comment by blog2read** [comment]
Would it be possilbe that the editor in file manager supports *.XSL files as well?

**Comment by qooob** [comment]
Yes. I have made an edit for it to be included into default package.

Meanwhile, to enable it now, edit manually `/filemin/defaultacl` and add to _allowed_for_edit_:

`application-xslt+xml`

Afterwards, it will let you edit it.

**Comment by brianforan** [comment]
Would it be possible to, within the theme, hide certain files or folders from appearing to users? Or would this be something to talk to Real-Gecko about?

**Comment by qooob** [comment]
What do you mean hide from users? Maybe using permissions?

**Comment by qooob** [comment]
It sound to me like a good feature - "Hide all files that user can't read."

**Comment by brianforan** [comment]
Yeah. So my goal with webmin/usermin is to give end users the ability to manage a single or two file(s) with Usermin>Filemin so they can up/down that file without needing FileZilla or some other client. 

I was thinking to use a custom script in the theme editor to hide those rows from the html table but if there is a way to do it in permissions that would work too, because I want to hide files/folders like `.usermin` type folders or `.bashrc` type files and only show the file(s) that they should have access too. Does that make sense?

**Comment by qooob** [comment]
It does but what your trying to do is not designed for File Manager usage.

**Comment by FBachofner** [comment]
It might be great to be able to "edit" bookmarks too.  [ not just delete ]

Right now the only (not-so-obvious) option is to edit the .bookmarks file.  This took me a while to find as I have it at /root/.filemin  instead of /home/myusername/.filemin  . . . even though I am logged into Webmin as myusername

**Comment by qooob** [comment]
Alright, I will add this to to-do as well.

**Comment by qooob** [comment]
This [commit](https://github.com/qooob/authentic-theme/commit/04266d1af0612c38448ba62f787c67c3d48c4d9f) is delivering feature number **_2 - Movable and resizeable editor windows_**

Besides mentioned, the new update will let the user to automatically **cascade windows** by clicking the key next to the save button:

![screenshot from 2017-05-14 18-34-03](https://cloud.githubusercontent.com/assets/4426533/26035409/25ea3d9e-38d4-11e7-91b0-4ad052798855.png)

This button will only appear if there are more than two visible panels. Double-clicking this button will rearrange the windows order, while cascading.

Very comfortable and _most powerful_ now, only with _Webmin_. :wink: :muscle:

**Comment by 7starsone** [comment]
👍 great... until now I had to open more browser tabs and now not needed anymore 🚀 👏 😄

**Comment by 7starsone** [comment]
would be possible to refresh a file content without closing it ? eg. put the case you want to refresh a log file from file manager instead of the usual tail on cli...
😉

**Comment by qooob** [comment]
Sure, will add it tomorrow. 😉

**Comment by 7starsone** [comment]
ok! can't wait... 👍 😃 thank you, bye

**Comment by 7starsone** [comment]
ops, forgot... would be also great to save "open files config"... eg. you often need x files opened, so you save those files and open all these files with a click on a button...something like this, if possible 😉

**Comment by qooob** [comment]
@7starsone I didn't understand your last message? Could you try again?

**Comment by 7starsone** [comment]
you often need some files opened:
eg.
`file1.txt`
`logfile1`
`logfile2`
...
and so on
so, put it in something like an "open batch file/button"
click on that button to open all those files at a time

like a "files list bookmark"
is that clear now?

**Comment by qooob** [comment]
You mean like a project in IDEs? 😁

**Comment by qooob** [comment]
Added _dropdown_ menu with yummy stuff for the editor.

1. Align Panels Vertically (when the menu is opened hold `Alt` to see the magic happening. Besides holding `Shift` will _reverse_ panels arrangement)
2. Toggle Soft Wrap
3. Refresh Content
4. Help (moved away from the top)

Is there anything else you want, folks?

![screenshot from 2017-05-16 17-03-14](https://cloud.githubusercontent.com/assets/4426533/26110124/0aee1fce-3a5a-11e7-9a9b-67e6f0c316fc.png)

**FIXED:** Multiple bugs with missing scrolling on collapsed panels, panel focus on clicking it, stand-alone window support and more..

**Comment by qooob** [comment]
https://github.com/qooob/authentic-theme/commit/f086f2ad9b8bfb2e567c2706d5ed7bd313888425

**Comment by rostovtsev** [comment]
Added support for creating symbolic links in File Manager.

![screenshot from 2018-10-03 15-08-32](https://user-images.githubusercontent.com/4426533/46409785-6a87d300-c71f-11e8-96d2-d589d935e57c.png)

**Comment by ilinbxu** [comment]
Google Chrome can't load file manager. demo：[https://www.scall.xyz/file.mp4](https://www.scall.xyz/file.mp4) File management is always loading.

The Firefox browser loads the file manager normally.

Sometimes it is normal, very confused.

**Comment by rostovtsev** [comment]
Hi,

May be it's done fancy cache issue? Hit `Ctrl+Alt(Options)+Shift+Meta(Control)+R`.

Is Firefox and Chrome on the same machine? Is the accessed server the same?

**Comment by rostovtsev** [comment]
Did it solve your issue?

**Comment by sz00gun** [comment]
If possible could you please fix better view on mobiles. Either menu more narrow or higher z-index for menu.
![Schowek02](https://user-images.githubusercontent.com/7549566/57732611-63ea1280-7695-11e9-9ce4-56bc8e1ba839.jpg)

That's not bad:
![Schowek03](https://user-images.githubusercontent.com/7549566/57732649-782e0f80-7695-11e9-8aaa-0fbbbcfd8f8f.jpg)

**Comment by leijonathan** [comment]
Not sure if this is the right place, but I would like an option to edit file manager settings on filemin. It would be useful to be able to enable/disable the encrypt feature and some other features for the virtual server owners.

**Comment by rostovtsev** [comment]
@leijonathan It's better to open separate issue.

I don't understand what you mean. What do you mean enable/disable encrypt feature?

**Comment by ochbob** [comment]
Hello @rostovtsev 

do you have any ETA about the point number 24 ?

Thank you very much.

**Comment by rostovtsev** [comment]
I am closing 19.3x branch soon with small bug fixes, and starting 19.4x. There I am adding new mail compose feature, and most likely, new read mail feature and other features, among which could be ACLs for File Manager.

Do you find this feature useful and worth time investments? I personally never had a need to deal with it. What are your real-life examples, when you would need it (just curious). I suppose it could go under existing Permissions modal, in case `getfacl/setfacl` commands are available and kernel supports it. Does device (fs) needs special `acl` flag upon mounting or it would just work?

**Comment by ochbob** [comment]
To be honest, I do not use ACL each day, but when I am using it, I prefer a graphical management, it's more clear, but i can deal with CLI.
I use ACL to manage user acces to specific files/folder. 
For example many user can have RW acces to specific folder but not the other one.

But if it's not a hard work I would love see this improvement to your file manager :D

If i understand your last sentence, I think it's specific to each fs.
For example, like ZFS, I should specified the correct flag to use ACL. (zfs set acltype=posixacl <dataset>)

**Comment by rostovtsev** [comment]
@ochbob Okay, I'll look into it for 19.40 release. :smiley_cat:

**Comment by snickers2k** [comment]
+1 for ACL

two other things would be

- small view/dropdown to monitor ongoing operations like file-transfers etc.
- split/dual-view

**Comment by ilinbxu** [comment]
> Hi,
> 
> May be it's sone fancy cache issue? Hit `Ctrl+Alt(Options)+Shift+Meta(Control)+R`.
> 
> Is Firefox and Chrome on the same machine? Is the accessed server the same?

Sorry, I just noticed the news, it is estimated that it is my local country network problem.

**Comment by rostovtsev** [comment]
@ilinbxu Which country in particular is that?

**Comment by ilinbxu** [comment]
In China, after I opened the VPN, it was normal.

**Comment by rostovtsev** [comment]
@ilinbxu Sorry to hear that, and I'm glad you're free from censorship now.

**Comment by zuhairmahd** [comment]
@rostovtsev it would be great if keyboard navigation support could be added to the file manager tree -- this would be very helpful for accessibility users (keyboard only users and screen reader users) who do not use a mouse.  Currently, the tree can receive focus, but you are unable to expand and collapse the tree using keystrokes, and also cannot navigate the tree.  Expected behavior is:
1. Up/down arrow: vertically navigate the tree
2. Left arrow: collapse the node your highlight is in
3. Expand the node your highlight is on

Note by the way that the screen reader can properly announce the name and collapse/expand status of the node receiving focus.

Please let me know if you'd rather I open a seperate issue and link it to this one -- would be glad to do that.

**Comment by rostovtsev** [comment]
Thank you for your suggestion. I will try to make it possible on the next release.

**Comment by rostovtsev** [comment]
Zuhair,

I have added this for you promptly. Try installing latest development version of the theme (19.40-alpha2) to give it a try. :smiley_cat:

**Comment by zuhairmahd** [comment]
@rostovtsev that was quick! It is working as expected.  I expected there to be a lag when moving in the tree while the other pain refreshes but it is actually pretty quick!

Thanks!

**Comment by rostovtsev** [comment]
> Is there a reason why double clicking on a file (e.g. public_html/index.html) does not open it or enable a preview / view of it ?

It sounds like possible improvement. I could do it, I think.

> Same for right click on file - why do we not have "view" or "preview" option? Only edit

Nobody ever asked for it, so I have never done it. I think it's worth adding it, you're right.

> Could it be possible to add drag / drop from right pane to the left main tree?

Umm, I guess it's possible. It will require some work though. Although, isn't it easier, and _safer_ using hotkeys, as _Ctrl-C/Ctrl-V_ (Cmd-C/Cmd-V on Mac) to copy/paste?

**Comment by rostovtsev** [comment]
I think the reason why I didn't make yellow icon on the tree, is that it would make the tree look clogged.

**Comment by rostovtsev** [comment]
> If possible could you please fix better view on mobiles. Either menu more narrow or higher z-index for menu.

Yes, it's better to use flexbox for it. Latest commit adds it. Have a look.

![image](https://user-images.githubusercontent.com/4426533/68947938-d0baa280-07c7-11ea-9953-e8fd0ed0b445.png)


The more should be done in terms of File Manager controls itself.

**Comment by rostovtsev** [comment]
Okay, it's better to use flexbox for the whole header. I dropped old table layout and switched to native Bootstrap rows and columns.

**Example:**


![flex_box_header](https://user-images.githubusercontent.com/4426533/68965347-a5967a00-07ec-11ea-8b41-9a24054fcba7.gif)

**Comment by campones** [comment]
One thing I'd like is when you hit ctrl+F to search, to be able to search the next code/word using the arrow, or by hitting a search button several times.   I have seen the scroll bar with the search result, it's nicely done

**Comment by rostovtsev** [comment]
> One thing I'd like is when you hit ctrl+F to search, to be able to search the next code/word using the arrow, or by hitting a search button several times.

You can already do it by hitting _F3_ to search forward and _Shift+F3_ to search back up. It's quite standard, even regular browsers support it nowadays.

>  I have seen the scroll bar with the search result, it's nicely done

Thanks. :smiley:

**Comment by campones** [comment]
Works great indeed, thanks!

**Comment by campones** [comment]
the f3 remembers me the search in the windows registry :)

**Comment by campones** [comment]
One cool thing would be when you create a file, you have the filename, as it is now, and the editor block   immediately below . instead to create and name the file, then click again on edit it

**Comment by rostovtsev** [comment]
You mean the newly created object (file or directory) would be auto-focused with the cursor?

**Comment by campones** [comment]
![image](https://user-images.githubusercontent.com/7072524/69757120-ce136200-1164-11ea-83ce-a616d373cd43.png)

**Comment by campones** [comment]
of course with your wonderful editor's formatting, and lines number (really missing in the old java file manager)

**Comment by rostovtsev** [comment]
It's possible, along with adding:

 1. Focus after create new file/folder/rename on the active row
 2. Focus nearby element when deleting
 3. Opening newly created file automatically.


Working on it. It will all be the part of 19.43 release super soon.

**Comment by rostovtsev** [comment]
The latest commit adds it and few more improvements. Try it with the latest dev version.

**Add ability to focus rows in post-actions in File Manager**

  1. After file/directory was deleted, the focus will go either to next or previous row
  2. After file/directory was renamed/created the focus will be placed on this row
  3. If file (not directory) was created, the editor will be opened automatically for it

**Fixed:**

  1. Calculation of file size for files containing HTML
  2. Fix (just a minutes ago) pagination calculations on post-actions

**Comment by campones** [comment]
I couldn't try last time I checked because I need a dev version of webmin 1.9.32 I believe.

**Comment by rostovtsev** [comment]
Well, there is a small mismatch in filesize labels but as far as I remember, it will not fail and will work just fine. You could force install it. You can always downgrade the theme to the needed version, using theme update script from its directory.

**Comment by campones** [comment]
ok , I ll try , it's a local server anyway. thank you

**Comment by campones** [comment]
I saw what you did, it's slightly different of what I thought,  you create the file and auto open the created file to insert the code. fine, it's a way of doing it.

since I forced the update, I have a weird problem, when I m deep down on the tree of file, if I right click a file it scrolls up alone tp the top

[![Screenshot from Gyazo](https://gyazo.com/41b31e929e5085742c6c8cbad348ff9a/raw)](https://gyazo.com/41b31e929e5085742c6c8cbad348ff9a)

**Comment by rostovtsev** [comment]
Thanks for the feedback.

I'll take a look tomorrow. It's super late right now.

**Comment by rostovtsev** [comment]
I have to do release now (_19.43_). The issue you are talking about will be addressed in _19.43-1_, just few days later.

**Comment by campones** [comment]
ok

2 notes I have: 

I noticed the nice homepage stats, it looks great but it uses quite some cpu 

also it could be nice to have either a word wrap in the editor, or even better, to be able to scroll horizontally with the mouse (but that would conflict with vertical scrolling if necessary so..)

**Comment by rostovtsev** [comment]
There is word-wrap already, just hit _Ctrl+Alt+W_ to toggle it or use top right drop-down. By the way, starting from version 19.43 the word-wrap option will be automatically remembered.

About CPU usage, how much resources does it use? How did you get the data?

**Comment by campones** [comment]
oh I see the wrap menu item indeed, very nice, that's a game changer compared to the java editor.. 

[![Screenshot from Gyazo](https://gyazo.com/3fbd058d458b6dc3e46fcd2a9e2f3857/raw)](https://gyazo.com/3fbd058d458b6dc3e46fcd2a9e2f3857)

can hit until 20% time to time

**Comment by rostovtsev** [comment]
Thanks.

You can disable Real-time stats in theme configuration, to decrease the load.

**Comment by campones** [comment]
yes, I did, 
did also install webmin 32 along with a reinstall of the theme 32 and no more issue with the right click on file

**Comment by rostovtsev** [comment]
You're right, it didn't work. I at first made this option just to be hostname based and later, hostname/logged in user based and there was a small bug.

Fixed now. Try again please. :)

**Comment by rostovtsev** [comment]
~~You're right, it didn't work. I at first made this option just to be hostname based and later, hostname/logged in user based and there was a small bug.

Fixed now. Try again please. :) ~~

Wrong thread, sorry!

**Comment by campones** [comment]
something that could be interesting would be to copy a path in the breadcrumb path and access it 

[![Screenshot from Gyazo](https://gyazo.com/deea53e6b58ad8807bb43aba2d64f5b0/raw)](https://gyazo.com/deea53e6b58ad8807bb43aba2d64f5b0)

**Comment by rostovtsev** [comment]
@campones  I specifically addressed this issue on the previous (19.43) release, as before it didn't work, with mouse click on the button, and only worked, if you hit enter key to submit. Right now, you can do both, what's more important you can paste any valid path in there, even with fancy chars.

Are you sure that you're running latest theme (19.43) version and don't have any stuck cache or anything?

**Comment by campones** [comment]
as I said, I downgraded to webmin/authentic 32 because of the previous issue while right clicking, and afaik, there is no webmin 33 deb file so I thought it was be wiser to have same version of webmin and theme

**Comment by rostovtsev** [comment]
I will take a look at this issue with jumping page on context menu, it's not critical but has to be fixed, I agree.

**Comment by campones** [comment]
well it makes the file manager very unpleasant. not sure if the reason is I have webmin 32 and it's not compatible with the theme v33. 

do you think it would be possible to make the whole area clickable?  so then it would really behave like browser tabs

**Comment by rostovtsev** [comment]
>  v33.
do you think it would be possible to make the whole area clickable? so then it would really behave like browser tabs

Do you mean, when the tab is collapsed, clicking all-over the header of collapsed tab, should normalise it back? (not just clicking an icon on the right)

**Comment by campones** [comment]
exactly

**Comment by rostovtsev** [comment]
Okay, I can do it, along with few other improvements, starting tomorrow.

**Comment by campones** [comment]
my feeling about the drop down menu item on the right top head is that it's good for mobile use, but unnecessary for desktop.   
The word wrap button, refresh button should always appear on the bar, the "save as" should be a sub menu item of the save button (adding drop down menu to the save button:   main icon (save) > drop down:  save as / save and close). 
the goal is to make as many possible with one click . 

I think also that there should be some distance between the typical editor functions (word wrap, utf, etc) and the tab management icons  "minimize, close and expand".

**Comment by rostovtsev** [comment]
I have addressed all mentioned issues on the latest commits. Thank you for your collaboration! :smiley_cat: 

About the toolbar - you can disable it in module configuration (cog icon on the top-right) by setting _Hide toolbar_ option to _Yes_.

**Comment by campones** [comment]
Ok. I ll give it a try a bit later

**Comment by campones** [comment]
I did force upgrade yesterday and so far, it's working fine,  big up for the cookie wordwrap.  

the tabs at the bottom now open by clicking anywhere on the top, as we discussed, but they also disappear from the "taskbar".  I think it would be cool if the tab was still present at the bottom independently the tab is minimized or not.  It would therefore act as any tab in a computer, and clicking again on the header would close/open it.  Well it's just an idea anyway.

**Comment by campones** [comment]
I cannot open an m3u8 file, which is a bit weird because it's a plain text file. 

also I don't understand how to change the icon, for .mp4 it shows "unknown"  while it could be multimedia or something

**Comment by rostovtsev** [comment]
Which server OS you are running there?

**Comment by campones** [comment]
ubuntu 14 and 18

**Comment by rostovtsev** [comment]
Make sure to have `shared-mime-info` package installed.

**Comment by campones** [comment]
> shared-mime-info is already the newest version.>

**Comment by campones** [comment]
Hi there, hope you are fine

look! today I was on the file manager and I got surprised that I couldn't find a top checkbox to select all files in a click . that good be a great addition, for mass delete for example

**Comment by iliarostovtsev** [comment]
There is a check mark icon already for doing it? Besides, you could do it by clicking * or + on your keyboard. Additionally, you can choose a first item, hold shift and click on the other item to select a range of files.

**Comment by campones** [comment]
oh I saw it now. I think it would be more intuitive if it was next to the icon folder to go up in the folder tree

https://i.gyazo.com/aec220e9dd241b723c272fa44ec1d6f2.mp4

btw you see my .ts file still have no icons

**Comment by bobrain008** [comment]
I suggest another feature: _Save Workspace_.
That means, put the case you want some logs files already opened with a click. At first you open all the text files you want or manually add to a list by file path, then **save the workspace** and assign any name for reference. Next time you open all the files with only one click.

**Comment by stevezoom** [comment]
My other host provides a "view" file option, which opens a window where editing is disabled. This mode prevents unintended changes while facilitating easy inspection of contents. Is that something you guys might consider in a future update?

**Comment by iliarostovtsev** [comment]
Yes, adding view option is a good idea. I'll add this to my todo.

**Comment by iliarostovtsev** [comment]
> My other host provides a "view" file option, which opens a window where editing is disabled. 

I am looking into adding **"View"** option. It would be also nice to have a quick preview of all files, including binary. The only possible tricky problem here is very large files, which should be loaded in chunks, not to crash the browser. If current API allows it, and it won't take way much time, I will do it, otherwise we can just limit the binary file to a certain number of lines on a preview. Another cool feature to have would be a _hex editor_ and _diff compare_ (for _diff compare_,  I was going to add it a long time ago but always had other things on the project to take care of.

I will work on adding **"View"** option to File Manager tomorrow and later ACLs as well, before making a new release of theme 19.70.

**Comment by iliarostovtsev** [comment]
> My other host provides a "view" file option, which opens a window where editing is disabled. This mode prevents unintended changes while facilitating easy inspection of contents. Is that something you guys might consider in a future update?

Okay, latest commit adds first step in previewing any files in File Manager. There are couple of important things missing but it can be tried already.

![image](https://user-images.githubusercontent.com/4426533/101684243-27e88080-3a77-11eb-8333-153e0ed2c5e4.png)


![image](https://user-images.githubusercontent.com/4426533/101684294-3afb5080-3a77-11eb-98c7-bc0ae7100ccd.png)


_Warning!_ At the moment, if you open super large files there are no truncation done on the server side and no loader shown (so it will take some time and might crash your browser's tab. It will be all fixed within next few days with step 2/2).

**Comment by iliarostovtsev** [comment]
Okay, final support for file viewer has been implemented. Now you can view any file with any size (requires latest patches for Webmin as well):

  1. Any file can be viewed; for binary files encoding and code highlight select are disabled
  2. Viewed file can be added to favorites
  3. Viewed file can be displayed in 5 different modes:
![image](https://user-images.githubusercontent.com/4426533/102720301-91cd1980-4304-11eb-96fe-e1aabbadb5c0.png)
  Default is showing 512 kb of a head of a file
  4. Opened file can be can be easily updated using `F2` or `Alt+R` hotkeys
  5. Both file editor and file viewer have now a footer for displaying status messages on the left and file size and status on the right:
![image](https://user-images.githubusercontent.com/4426533/102720479-a8c03b80-4305-11eb-9c8b-05493a01fa2c.png)


**Tip:** Try adding a log file with mode _Tail and reverse file output_, while using `F2` for getting changes updates instantly ..

**Comment by iliarostovtsev** [comment]
Forgot to mention: 

  1. Files can be viewed from command shell, by typing `view /etc/hosts` or straight from nav menu `!view /etc/hosts`
  2. While in File Manager clicking `F3` will invoke viewer window with default preview options (showing first 512 kb of a head of a file)

**Comment by campones** [comment]
Hey, could a mp4 player be added to preview mp4 files?

**Comment by iliarostovtsev** [comment]
Yeah, a we have a picture preview already, adding a video preview would make sense. I will add this to my todo.

**Comment by campones** [comment]
I would use videojs

**Comment by iliarostovtsev** [comment]
> I would use videojs

I will take a look. I remember I used _mediaelementjs_ a long time ago for my `karateka.org` project. Perhaps _video.js_ would be a better fit.

**Comment by iliarostovtsev** [comment]
.. and perhaps nothing at all would be needed for a single video preview ..

**Comment by campones** [comment]
well videojs is actively maintained, not sure about mediaelement

**Comment by sz00gun** [comment]
> well videojs is actively maintained, not sure about mediaelement

You won't find anything better than video.js

**Comment by iliarostovtsev** [comment]
> You won't find anything better than video.js

Okay, thanks, I will have a look and see if it fits. I am going to implement this the same way we do with photos, by gathering all videos from a directory and putting it into a playlist.

**Comment by iliarostovtsev** [comment]
The latest commit improves the search capability done using File Manager:

What was added for search:

  1. Ability to limit to user and group while searching
  2. Limit file size - you can use comparison operators, like `<` or `>=` or `!=` or `==` as well as units like `KiB` or `Mb` or `T` for example to search only files larger than 1 megabyte you can go with `>1M` or `>1 Mb` or `>1MiB` or `> 1024`
  3. Limit file type to _File_, _Directory_, _Symlink_ - if directory is selected the file size filter will work on directories too!

**Example:**
![image](https://user-images.githubusercontent.com/4426533/103096854-5c7f3f00-4616-11eb-9075-d0223f95911b.png)

**Comment by iliarostovtsev** [comment]
I also added size range file search, which can be used to search files with size larger than and smaller than, like this:

Find all files larger than 3 MiB and smaller than 5 MiB:
~~~
>3M - <5M
~~~

**Comment by iliarostovtsev** [comment]
I have finished work today on the them **19.70-beta3** which adds a new cool feature to File Manager - ability to move files to _Recycle Bin_, rather than deleting completely.

**Explanations on how it works:**

  1. By default files are deleted as before (irreversibly - called _delete_ option)
  2. Files now can be moved to trash (called _remove_ option) by doing the following:
    2.1. Holding `Alt` key (Mac `Option` key) while context menu is opened
    2.2. Setting configuration option in module configuration called **Delete files safely by moving to the trash** to **Yes**.
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; _Note:_ if set to _Yes_, when holding `Alt` key, regular, irreversible delete option appear.

 
        ![image](https://user-images.githubusercontent.com/4426533/103386552-6ae6c280-4b10-11eb-9177-d7fd2fa844e0.png) ![image](https://user-images.githubusercontent.com/4426533/103386531-4ee32100-4b10-11eb-93d9-6efd189a9175.png)

  3. Dialog window title, icon and button text, which pops up upon deletion will depend on the mode:
![image](https://user-images.githubusercontent.com/4426533/103386635-e5174700-4b10-11eb-90a8-415006f94c05.png) ![image](https://user-images.githubusercontent.com/4426533/103386649-f2343600-4b10-11eb-8a35-95007f7a6f01.png)

  4. Deleted files when moved to **Trash** will be placed under user home directory, i.e. **~/.Trash** . Permissions on targeted files and directories will be saved. If there is a clash on directories or files found they will be appended with datetime string, like this:
~~~
/root/.Trash/home/ubuntu20-pro/public_html-2020-12-31_23:59:59
~~~

Looking forward for your suggestions and/or comments.

---

I wish everyone to have a great holidays and a happy new year!

**Comment by sz00gun** [comment]
Wow! Amazing facility, however is possible to locate the .Trash folder in /home instead of /root ?

/root in my case don't have much left space, but /home have a lot, because there are all website.

**Comment by iliarostovtsev** [comment]
> Wow! Amazing facility

Good to hear! :slightly_smiling_face: 

> however is possible to locate the .Trash folder in /home instead of /root ?

Well, if files are deleted by the user (server owner) then they will be placed under `/home`, while if `root` user is deleting files, then files will be placed to `/root` because it defaults to home directory. Users without home will use `/` instead.

What I could theoretically do is to move files to `/home/username/.Trash`, if files were deleted from virtual server's home by `root` user (master administrator).

To summarize, right now, it's user based only but I could make it home dependable for virtual servers, the same way we automatically assign ownership to files created by `root` under virtual server's home.

Honestly I had this idea but I wasn't sure if files deleted by `root` should be moved under virtual server's home directory instead.

Perhaps adding another option (oh no) to File Manager configuration page, which would let define a single trash directory for all users? Then one could mount a super large drive to `/mnt/.Trashed` and collect trash for as long as needed?

**Comment by campones** [comment]
great work as usual, happy new year!

**Comment by mitchellkrogza** [comment]
> > Wow! Amazing facility
> 
> Good to hear! 
> 
> > however is possible to locate the .Trash folder in /home instead of /root ?
> 
> Well, if files are deleted by the user (server owner) then they will be placed under `/home`, while if `root` user is deleting files, then files will be placed to `/root` because it defaults to home directory. Users without home will use `/` instead.
> 
> What I could theoretically do is to move files to `/home/username/.Trash`, if files were deleted from virtual server's home by `root` user (master administrator).
> 
> To summarize, right now, it's user based only but I could make it home dependable for virtual servers, the same way we automatically assign ownership to files created by `root` under virtual server's home.
> 
> Honestly I had this idea but I wasn't sure if files deleted by `root` should be moved under virtual server's home directory instead.
> 
> Perhaps adding another option (oh no) to File Manager configuration page, which would let define a single trash directory for all users? Then one could mount a super large drive to `/mnt/.Trashed` and collect trash for as long as needed?

You are a STAR !!!! Thank you for adding this suggestion I sent you merely 3 days ago

**Comment by iliarostovtsev** [comment]
Upcoming Authentic Theme release will add support for displaying detailed file info, by clicking **`Alt+Enter`** or **`Command + I`** on a file while using keyboard navigation or calling it from the context menu. File checksums will also be displayed.

**Example:**

![image](https://user-images.githubusercontent.com/4426533/120231840-edf47c00-c25a-11eb-81b5-03c0426eeefd.png)
![image](https://user-images.githubusercontent.com/4426533/120231997-4297f700-c25b-11eb-8789-be7a8b8d8240.png)

![image](https://user-images.githubusercontent.com/4426533/120231929-18ded000-c25b-11eb-8554-8a35fc9cb5c9.png)

**Note:** I am looking extending this output with ACLs as well (when supported).

---

If there is anything missing or you wish to see any extra fields on the output that I missed, please let me know.

**Comment by Sopor** [comment]
When sorting by size the directories will also be sorted by size even if it doesn't know the size of the directories.

I would like to have an option that always keep the directories at top and if it knows the size of the directories it should sort them too, but they should still be at the top.

**Comment by iliarostovtsev** [comment]
Good suggestion. Will do.

**Comment by iliarostovtsev** [comment]
Fixed.

**Comment by sz00gun** [comment]
A very little problem with pop-ups in file manager on mobile phones. Eg. (?) Button and (bookmarks) button. 
![IMG_20211210_113436](https://user-images.githubusercontent.com/7549566/145568489-ce996036-2988-43e6-9dc3-66d4355589b1.jpg)

**Comment by iliarostovtsev** [comment]
Duh. Really annoying! Thanks for reporting. I'll have a look.

**Comment by WolfgangDpunkt** [comment]
> 2\. Files now can be moved to trash (called _remove_ option)

Is the recycle bin automatically emptied on a regular basis? This would be a nice feature.
Or does it have to be done manually?

I found the command in the context menu of the trash folder, maybe an improvment would be to include it centrally at the menu bar, for example within _Tools_? "Empty all recycle bins" or, for risk-averse users: "Purge all data in the recycle bin that is older than one month".

Thank you very much for this great software.

**Comment by iliarostovtsev** [comment]
> Thank you very much for this great software.

Very welcome! 🙂 

> Is the recycle bin automatically emptied on a regular basis? This would be a nice feature. Or does it have to be done manually?

At the moment you would have to do it manually but adding a button to do it would probably be a good feature. Although it's not clear what root capable user action would be like -- would you expect it to be deleted for all users (like under `/home`)
 or only for its own home dir?

Also, latest version can do purging trash by right clicking on it and choosing **Empty Trash** from context menu:

![image](https://user-images.githubusercontent.com/4426533/145995189-a795c84b-2019-428b-892c-44f13395f95b.png)

**Comment by WolfgangDpunkt** [comment]
I really appreciate your quick reply.

> Also, latest version can do purging trash by right clicking on it and choosing **Empty Trash** from context menu

I had actually found that earlier, but more intuitive would probably really be a menu button.

I use Webmin as only user and admin on my RPI4, so I'm not representative for an multi-user perspective.
Maybe two options then? "Purge my trash" and "Purge all trash".

I personally will probably than use for automation purging simply a monthly cron task.

**Comment by iliarostovtsev** [comment]
> Maybe two options then? "Purge my trash" and "Purge all trash".

Good suggestion. I'll think about adding it.

> I personally will probably than use for automation purging simply a monthly cron task.

Yeah, this will work too.

**Comment by iliarostovtsev** [comment]
> @sz00gun A very little problem with pop-ups in file manager on mobile phones. Eg. (?) Button and (bookmarks) button.

This is fixed. Thanks for reporting.

**Comment by iliajie** [comment]
> I found the command in the context menu of the trash folder, maybe an improvment would be to include it centrally at the menu bar, for example within _Tools_? "Empty all recycle bins" or, for risk-averse users: "Purge all data in the recycle bin that is older than one month".

This feature will be implemented in upcoming Authentic Theme 19.91.1 release.

When move to trash option is activated it will be possible for master administrator to purge all trash directories for all users or only own trash directory:

<img width="792" alt="image" src="https://user-images.githubusercontent.com/4426533/160498097-cbd0ed1e-eb5a-44ac-a450-e2fe5f9dec50.png">

**Comment by Mecha-Weasel** [comment]
Drooling over #13 "Edit/reorder bookmarks (not just delete/add)"

**Comment by iliajie** [comment]
I will look into adding it before a new theme release.

**Comment by ochbob** [comment]
Hello @iliajie 

Any news about #23 ? 

thank you !

**Comment by iliajie** [comment]
> Any news about #23 ?

Never had a moment to implement ACLs controls. Although, listing been added a while ago and can be enabled in File Manager configuration.

**Comment by b-sullender** [comment]
When editing a text file using Webmin’s File Manager on an iPad the focus of input goes to the content filter, preventing me from entering text into the file. This might technically be a bug, but the option to turn off the content filter might be an easy and fast fix that adds more functionality.

**Comment by iliajie** [comment]
Thanks for the heads up. This issue has been fixed on the latest theme 20.22.

**Comment by iliaross** [comment]
> Add ACLs control Issue with File ALCs module #1229 (comment)

Those were added a while ago.
